Process channels with actor model

contrun opened the pull request.

contrun committed:
Add placeholding code to update channels

contrun committed:
Add various types for channels

contrun committed:
stash

contrun committed:
Handle OpenChannel message

contrun committed:
Add skeleton code to handle AcceptChannel

contrun committed:
Use a unified result type for processing channel logic

contrun committed:
Derive the same channel id

contrun committed:
Remove channel manager

contrun committed:
Add more fields to open and accept channels

contrun committed:
Add function to get channel parameter to create transactions

contrun committed:
Fix deriving private keys in each round

contrun committed:
Check consistency of deriving private and public keys

contrun committed:
Wrap secp256k1 crate instead of ckb_crypto

contrun committed:
Wrap Byte32 in molecule instead of directly using it

contrun committed:
Add test for serde public key

contrun committed:
Refactor key derivation while opening channel

contrun committed:
Add field next_local_nonce to OpenChannel

contrun committed:
Respond AcceptChannel message automaticall on receiving OpenChannel

contrun committed:
Add field next_local_nonce to AcceptChannel

contrun committed:
Fix sending invalid AcceptChannel message

contrun committed:
Fix serialization of next_local_nonce in {Accept,Open}Channel

contrun committed:
Save channel while opening channel

contrun committed:
Move e2e test for channel open use and closure to dedicated folder

contrun committed:
Use ractor

contrun committed:
Initialize NetworkControllerActor in pre_start

contrun committed:
Create PeerActor to process peer related logic

contrun committed:
Split ckb network functionality into serveral actors

contrun committed:
Further modularize different actors

contrun committed:
Implement more functions in actors

contrun committed:
Fix tentacle not gracefully shut down on SIGTERM received

contrun committed:
Keep peer state in network actor

contrun committed:
Add NetworkServiceEvent to network actor

contrun committed:
Remove field event sender from tentacle Handle

contrun committed:
Add an actor to process events

contrun committed:
Add an ancestry actor to control all other actors' life time

contrun committed:
Remove separated events processing logic

contrun committed:
Chore: remove a few unused things

contrun committed:
Don't return ActorProcessingError

Because ractor considers this a fatal error and exits the runtime.

contrun committed:
Don't panic on network actor shutdown before tentacle service

contrun committed:
Set response in rpc

contrun committed:
Demonstrate who is blocking

contrun committed:
Revert "Demonstrate who is blocking"

This reverts commit b98ac25e1e8ea847b86f1a5a334b2062229c6921.

contrun committed:
Use tokio channel instead of oneshot to avoid blocking

contrun committed:
Encapsulate sending actor message in tenatcle handler

contrun committed:
Remove some too verbose debugging messages

contrun committed:
Add unit test for connecting a non existent peer

contrun committed:
Sleep for sometime after dialing a peer

contrun committed:
Fix funding amount mismatch

contrun committed:
Fix state not update while OpenChannel message sent

contrun committed:
Add a note on funding_amount

contrun committed:
Add handling of transaction collaboration messages

contrun committed:
Simplify network handling by removing PeerActor

contrun committed:
Add Command varint to ChannelActor message

contrun committed:
Refactor to make it easier to add channel commands

contrun committed:
Add channel command for transaction collaboration

contrun committed:
Remove outdated command.rs

contrun committed:
Add github actions workflow file

contrun committed:
Simplify channel message handling

contrun committed:
Implement updating channel id while handling accepting channels

doitian committed:
rpc: split rpc into nested routers

contrun committed:
Remove transaction while processing message

contrun committed:
Handle remote commitment signed message

contrun committed:
Handle local commitment signed command

contrun committed:
Add skeleton for generating and verifying partial signature

contrun committed:
Add a few tx_add messages

contrun committed:
Fix channel id not correctly set in acceptor

contrun committed:
Fix state not updated while opening/accepting channel

contrun committed:
Fix incorrect handling of tx collaboration finished earlier

contrun committed:
Sleep sometime to make processing sequential

contrun committed:
Overhaul tx collaboration logic

contrun committed:
Overhaul commitment signed logic

contrun committed:
Order nonces in musig2

doitian commented on src/ckb/schema/pcn.mol:
```diff
@@ -1,6 +1,7 @@
 
-array Signature [byte; 65];
+array Signature [byte; 64];
+array Byte66 [byte; 66];nonce
```
```suggestion
array Byte66 [byte; 66]; // nonce
```

contrun replied:
Byte66 is changed to PubNonce. I currently serialize Signature to the compact form. So it is 64 bytes. I don't fully understand the implication of using 65 bytes form of signature vs 64 bytes of signature. Can you elaborate on that?

doitian commented on src/ckb/types.rs:
```diff
@@ -1,88 +1,218 @@
+
+impl From<&PubNonce> for Byte66 {
+    fn from(value: &PubNonce) -> Self {
+        Self::new_builder()
```
It's simpler to use `Byte66::from_slice`

contrun committed:
Move task related functionos to a module

quake committed:
chore: tweak debug level

quake committed:
fix: tx collaboration

contrun committed:
Add skelton code to build commitment transaction

contrun committed:
Add function to obtain channel paramters

contrun committed:
Reorganize impl ChannelStateActor for readability

contrun committed:
Remove a few unused types

contrun committed:
Fix using signature for funding tx in commitment signed

contrun committed:
Use the correct commitment point while funding channel

contrun committed:
Send another commitment_signed message

contrun committed:
Process tx_signatures message

contrun committed:
Process channel_ready message

contrun committed:
Add help function to create a network node

contrun committed:
Add option to keep temp directory for tests

contrun committed:
Add helper function to connect to peer

contrun committed:
Add helper function to create multiple connected nodes

contrun committed:
Add unit test to open an outbound channel

contrun committed:
Rename molecule type Byte66 to PubNonce

contrun committed:
Fix not shutting down tasks

contrun committed:
Cargo clippy

contrun committed:
Complete ChannelReady logic

contrun committed:
Tweak a few test helper functions

contrun committed:
Add dedicated LockTime type

contrun committed:
Handle add_tlc

contrun committed:
Fix not correctly set preimage in add_tlc

contrun committed:
Test add_tlc rpc requests

contrun committed:
Test add_tlc rpc requests

contrun committed:
Split AddTLCMessage and AddTLCCommand processing logic

contrun committed:
Handle TlcRemove

Commit:
funding: manage funding tx in an actor

quake committed:
refactor: change amount from u64 to u128

contrun committed:
Document the problem of state updates in NetworkActor and ChannelActor

contrun committed:
Fix typos and grammar errors

quake committed:
refactor: tx_update message

contrun committed:
Implement RPC to accept channels

contrun committed:
Update test cases for AcceptChannel RPC

contrun committed:
Abstract seed generation in network

contrun committed:
Update test case for new channel temp id generation method

contrun committed:
Fix out dated comments

contrun committed:
Add trailing node info for debugging

contrun committed:
Fix state not transitioned to AwaitingTxSignatures in some case

contrun committed:
Fix tx_signatures sending order

contrun committed:
Refine how to create a broadcastable commitment transaction

contrun committed:
Handle CommitmentSigned message while channel is ready

contrun committed:
Create more realistic tx inputs to funding tx

contrun committed:
Handle shutdown

contrun committed:
Add more commands in e2e tests

contrun committed:
Implement auto revoke and ack

contrun committed:
Implement shutdown command to send shutdown message

contrun committed:
Implement build closing transaction

contrun committed:
Handle ClosingSigned

contrun committed:
Add more information to generate commitment transaction correctly

contrun committed:
Save historical commitment points of counterparty

contrun committed:
All commitment numbers should start with zero

contrun committed:
Simplify commitment point and channel parameter obtaining

contrun committed:
Fix incorrect balance distribution while establishing channel

contrun committed:
Check the validity of revoke and ack message

contrun committed:
Refactor ordering things for musig2

contrun committed:
Fix {add,remove}-tlc

contrun committed:
Remove pending tlc before closing_signed

contrun committed:
Automatically send closing_signed message on all pending tlcs removed

contrun committed:
Pad zero to filenames

contrun committed:
Create commitment lock sucessfully

contrun committed:
Fix to broadcaster value

contrun committed:
Sort tlcs to generate the same commitment transaction

contrun committed:
Fix a few fields in generating commitment transaction

contrun committed:
Fix delay in commitment tx not correctly set

contrun committed:
Try to add a simple test

contrun committed:
Use my determinisitic for ckb-testtool

contrun committed:
Try to verify transactions

contrun committed:
Comment out funding tx verification

contrun committed:
Make sure we can revoke txs

contrun committed:
Try to run commitment lock on commitmeng signed

contrun committed:
make tests self contained

contrun committed:
Fix using incomplete celldeps

contrun committed:
Fix wrong script in commitment tx

contrun committed:
Use tx to spend commitment tx to verify commitment tx validity

contrun committed:
Add more logs to context.complete_tx

contrun committed:
Demonstrate complete_tx does something unexpected

contrun committed:
Fix script not found

contrun committed:
use rwlock instead of mutex for context

contrun committed:
Add an always_success script

contrun committed:
Fix script not found error

contrun committed:
Fix using the wrong index in creating transaction from commitment lock

contrun committed:
cargo fmt

contrun committed:
add a unit test to see if built funding tx works

contrun committed:
Fix commitment tx not saved

contrun committed:
Remove some duplicated code

contrun committed:
Use witness script from previous round instead of current one

contrun committed:
Use newest ckb-testtool on github

contrun committed:
Fix initial commitment number

contrun committed:
Don't update holder commitment number on sending CommitmentSigned

contrun committed:
Make ckb depend on ckb-chain service

contrun committed:
Add script to quickly depoly contracts to devchain

contrun committed:
Make a fortune for node 3 in test

    private key: d00c06bfd800d27397002dca6fb0993d5ba6399b4238b2f29ee9deb97593d2bc
    CKB Testnet address: ckt1qyqvsv5240xeh85wvnau2eky8pwrhh4jr8ts8vyj37

has many hard-coded balances

contrun committed:
Support {funding,commitment}-lock in deployment script

contrun committed:
Update binary directory

contrun committed:
Rename temp to chain

contrun committed:
Remove unused code in commitment lock context

contrun committed:
Better encapsulate commitment lock context in channels

contrun committed:
Fix not using the wrong commitment secret to verify RevokeAndAck

contrun committed:
Remove a few unused code

contrun committed:
Refactor chain in ckb so that we can easily add real chain config

contrun committed:
Simplify MockContext building

contrun committed:
Pass on chain scripts info on pre_start to CkbChainActor

contrun committed:
Use real chain config for contracts

contrun committed:
Fix tests

contrun committed:
Use real dev chain config

contrun committed:
Transfer some balance to node 1 on force rerun dev nodes

contrun committed:
Don't panic when build unlocked failed

contrun committed:
Minor fix

contrun committed:
Automatically fetch funding tx when channel accepted

contrun committed:
Refactor handling fund tx update for easier update

contrun committed:
Implement tx update workflow

contrun committed:
Use script for FundingRequest

contrun committed:
Fix infinite loop in tx collaboration

contrun committed:
Remove old worlflow of manually add tx inputs

contrun committed:
Create valid witness from ckb chain actor

contrun committed:
Broadcast funding tx

contrun committed:
Make a larger channel to avoid min occupied capacity problem
